fate/src/main/java/org/apache/accumulo/fate/util/Retry.java

```diff
@@ -0,0 +1,356 @@
+     */
+    NeedsTimeIncrement retryAfter(long duration, TimeUnit unit);
+  }
+
+  public interface NeedsTimeIncrement {
+    /**
+     * @param duration
+     *          the amount of additional time to add before each subsequent retry; input is converted to milliseconds, rounded down to the nearest
+     * @return this builder with the increment amount set
+     */
+    NeedsMaxWait incrementBy(long duration, TimeUnit unit);
+  }
+
+  public interface NeedsMaxWait {
+    /**
+     * @param duration
+     *          the maximum amount of time to which the waiting period between retries can be incremented; input is converted to milliseconds, rounded down to
+     *          the nearest
+     * @return this builder with a maximum time limit set
+     */
+    NeedsLogInterval maxWait(long duration, TimeUnit unit);
+  }
+
+  public interface NeedsLogInterval {
+    /**
+     * @param duration
+     *          the minimum time interval between logging that a retry is occurring; input is converted to milliseconds, rounded down to the nearest
+     * @return this builder with a logging interval set
+     */
+    BuilderDone logInterval(long duration, TimeUnit unit);
+  }
+
+  public interface BuilderDone {
+    /**
+     * Create a RetryFactory from this builder which can be used to create many Retry objects with the same settings.
+     *
+     * @return this builder as a factory; intermediate references to this builder cannot be used to change options after this has been called
+     */
+    RetryFactory createFactory();
+
+    /**
+     * Create a single Retry object with the currently configured builder settings.
+     *
+     * @return a retry object from this builder's settings
+     */
+    Retry createRetry();
+  }
+
+  public interface RetryFactory {
+    /**
+     * Create a single Retry object from this factory's settings.
+     *
+     * @return a retry object from this factory's settings
+     */
+    Retry createRetry();
+  }
+
+  public static NeedsRetries builder() {
+    return new RetryFactoryBuilder();
+  }
+
+  private static class RetryFactoryBuilder implements NeedsRetries, NeedsRetryDelay, NeedsTimeIncrement, NeedsMaxWait, NeedsLogInterval, BuilderDone,
+      RetryFactory {
+
+    private boolean modifiable = true;
+    private long maxRetries;
+    private long initialWait;
+    private long maxWait;
+    private long waitIncrement;
+    private long logInterval;
+
+    RetryFactoryBuilder() {}
+
+    private void checkState() {
+      Preconditions.checkState(modifiable, "Cannot modify this builder once 'createFactory()' has been called");
+    }
+
+    @Override
+    public NeedsRetryDelay infiniteRetries() {
+      checkState();
+      this.maxRetries = -1;
+      return this;
+    }
+
+    @Override
+    public NeedsRetryDelay maxRetries(long max) {
+      checkState();
+      Preconditions.checkArgument(max >= 0, "Maximum number of retries must not be negative");
```

Comment:
could add unit test for these new checks

Reply:
Wilco